resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/resources/regression/Bug_297635.java

```diff
@@ -0,0 +1,147 @@
+			try {
+				saveSnapshot(saveManagerSpy);
+			} catch (CoreException e) {
+			}
+			verify(saveManagerSpy).forgetSavedTree(saveParticipantBundle.getSymbolicName());
+		});
+	}
+
+	private void saveFull() throws CoreException {
+		getWorkspace().save(true, getMonitor());
+	}
+
+	private void saveSnapshot(SaveManager saveManager) throws CoreException {
+		saveManager.save(ISaveContext.SNAPSHOT, true, null, getMonitor());
+	}
+
+	private void executeWithSaveManagerSpy(Consumer<SaveManager> executeOnSpySaveManager)
+			throws NoSuchFieldException, SecurityException, IllegalArgumentException, IllegalAccessException {
+		IWorkspace workspace = getWorkspace();
+		String saveManagerFieldName = "saveManager";
+		SaveManager originalSaveManager = (SaveManager) getField(workspace, saveManagerFieldName);
+		SaveManager spySaveManager = spy(originalSaveManager);
+		try {
+			setField(workspace, saveManagerFieldName, spySaveManager);
+			executeOnSpySaveManager.accept(spySaveManager);
+		} finally {
+			setField(workspace, saveManagerFieldName, originalSaveManager);
+		}
+	}
+
+	private static Object getField(Object object, String fieldName)
+			throws NoSuchFieldException, SecurityException, IllegalArgumentException, IllegalAccessException {
+		Field field = object.getClass().getDeclaredField(fieldName);
+		field.setAccessible(true);
+		return field.get(object);
+	}
+
+	private static void setField(Object object, String fieldName, Object value)
+			throws NoSuchFieldException, SecurityException, IllegalArgumentException, IllegalAccessException {
+		Field field = object.getClass().getDeclaredField(fieldName);
+		field.setAccessible(true);
+		field.set(object, value);
+	}
+
+	private static class SaveParticipant implements ISaveParticipant {
+		private static final String SAVE_PARTICIPANT_BUNDLE = "org.eclipse.bundle01";
+
+		@Override
+		public void doneSaving(ISaveContext context) {
+			// nothing to do
+		}
+
+		@Override
+		public void prepareToSave(ISaveContext context) {
+			context.needDelta();
+			context.needSaveNumber();
+		}
+
+		@Override
+		public void rollback(ISaveContext context) {
+			// nothing to do
+		}
+
+		@Override
+		public void saving(ISaveContext context) {
+			// nothing to do
+		}
+
+		public static Bundle installContainingBundleAndRegister()
+				throws BundleException, MalformedURLException, IOException, CoreException {
+			Bundle bundle = BundleTestingHelper.installBundle("1", getContext(),
+					ContentTypeTest.TEST_FILES_ROOT + "content/bundle01");
+			BundleTestingHelper.resolveBundles(getContext(), new Bundle[] { bundle });
+			bundle.start(Bundle.START_TRANSIENT);
+			registerAtWorkspace();
+			return bundle;
+		}
+
+		private static BundleContext getContext() {
+			return Platform.getBundle(PI_RESOURCES_TESTS).getBundleContext();
+		}
+
+		private static void registerAtWorkspace() throws CoreException {
+			getWorkspace().addSaveParticipant(SAVE_PARTICIPANT_BUNDLE, new SaveParticipant());
+		}
```

Comment:
These 3 methods don't need to be in this class, you could move them to the outer class.

Reply:
Yes, they could technically be moved. But why should they? They semantically belong to the `SaveParticipant`, which is why they are placed in that class.

Reply:
> Instantiating, installing and registering is something that is done _on_ a `SaveParticipant` but not _by_ a `SaveParticipant`.

True, that's why it was no instance method but a class method, comparable to having a factory method for a class.
But I think I got your point and agree that the inner class was not well designed, so I restructured it in f9a5b76. Now it makes more sense to me and I hope you agree :-)

Reply:
I agree, thanks for changing it